tsdb/head.go

```diff
@@ -1622,15 +1624,57 @@ func (h *headIndexReader) SortedLabelValues(name string) ([]string, error) {
 
 // LabelValues returns label values present in the head for the
 // specific label name that are within the time range mint to maxt.
-func (h *headIndexReader) LabelValues(name string) ([]string, error) {
+// If matchers are specified the returned result set is reduced
+// to label values of metrics matching the matchers.
+func (h *headIndexReader) LabelValues(name string, matchers ...*labels.Matcher) ([]string, error) {
 	h.head.symMtx.RLock()
-	defer h.head.symMtx.RUnlock()
+
 	if h.maxt < h.head.MinTime() || h.mint > h.head.MaxTime() {
+		h.head.symMtx.RUnlock()
 		return []string{}, nil
 	}
 
-	values := h.head.postings.LabelValues(name)
-	return values, nil
+	if len(matchers) == 0 {
+		h.head.symMtx.RUnlock()
```

Comment:
I think that should still be deferred ?

Reply:
the reason why i changed it non-deferred is because that way we can release the lock earlier on `:1666` if matchers have been specified. 
there is a trade-off between the safety of using deferred unlocks (because no-one can forget to unlock) and reducing lock congestion by unlocking early when it is possible.
not sure which side of this trade-off would be preferable, since the TSDB is a pretty integral and optimized part of Prometheus in general I chose the latter, but I'm happy to change this back if you disagree.